File: backend/app/harness/agents/memory_markdown_agent.py
"""Memory Markdown Agent — 从对话增量提炼 .md 记忆。

两个入口：
- refine_user_markdown: 用户级（身份/职业/研究大方向/偏好）
- refine_project_markdown: 项目级（本项目研究方向/子问题/关注点）

设计要点：
- **绝不**自己编造用户没说过的内容；只整理/归并 LLM 从对话中看到的线索
- 保留用户已手写的 section；只对"空占位"或"已有条目"做增补
- 失败或 LLM 不可用时，静默返回 `current_markdown` 不动（前端按原值呈现）
"""
from __future__ import annotations

import logging
from typing import Optional

from app.services.core.llm_config_store import get_llm_manager

logger = logging.getLogger(__name__)


_USER_PROMPT_TEMPLATE = """你是一个记忆整理助手。阅读下方"当前记忆 Markdown"与"最近对话"，
只根据对话中用户明确说过的信息，补全或修订这份用户级 Markdown。

【硬性规则】
1. 仅使用对话中**用户本人**提到的身份/职业/研究大方向/偏好；**禁止编造**、
   禁止从"学科词汇"猜职业/机构，禁止引入与用户无关的研究主题。
2. 保留原 Markdown 的结构与用户已手写的内容；只在空占位或已有列表里增补。
3. 不要把"当前项目"里的具体研究方向（如某篇论文、某个子问题）写入用户级 —— 那是项目级的事。
4. 用户级只写稳定身份：昵称、性别（仅当明确）、年龄/年级（仅当明确）、职业/学生/研究员、
   所在机构、常用语言、研究大方向（如"计算机系统/机器学习/生物信息"）、数据源偏好。
5. 返回**完整**的新 Markdown（而非 diff），保持 UTF-8、中文书写、Markdown 语法合法。
6. 如果对话里没有任何新信息能增补，**原样返回** 当前 Markdown，一字不改。

【当前用户】{user_name}

【当前记忆 Markdown】
{current_markdown}

【最近对话】
{conversation}

【输出】
只输出新的 Markdown 本体，不要包含 ``` 代码块围栏、不要前后解释。
"""


_PROJECT_PROMPT_TEMPLATE = """你是一个项目研究记忆整理助手。阅读"当前项目记忆 Markdown"与"项目内对话"，
只根据对话中用户明确表达的研究方向/子问题/关注点，补全或修订这份项目级 Markdown。

【硬性规则】
1. 仅写对话中出现的本项目研究方向、核心子问题、关键术语、用户关注的文献/思路、近期关注点。
2. **禁止**引入对话中完全没出现的领域词（反污染：不要把别的学科词写进来）。
3. 保留用户手写部分；只在空占位或已有列表里增补。
4. 不要把用户身份/职业（那属于用户级）写入项目级。
5. 返回**完整**的新 Markdown。
6. 如无新信息，**原样返回**。

【项目标题】{project_title}

【当前项目记忆 Markdown】
{current_markdown}

【项目内对话】
{conversation}

【输出】
只输出新的 Markdown 本体，不要代码块围栏、不要前后解释。
"""


def _format_conversation(messages: list[dict]) -> str:
    lines: list[str] = []
    for m in messages:
        role = m.get("role", "user")
        tag = {"user": "用户", "assistant": "AI", "system": "系统"}.get(role, role)
        content = (m.get("content") or "").strip()
        if not content:
            continue
        lines.append(f"[{tag}] {content}")
    return "\n\n".join(lines) if lines else "(无对话)"


async def _call_llm(prompt: str) -> Optional[str]:
    try:
        llm = await get_llm_manager()
        result = await llm.generate(prompt, temperature=0.1, max_tokens=3000)
        return (result or "").strip() or None
    except Exception as e:
        logger.warning("memory_markdown_agent LLM call failed: %s", e)
        return None


async def refine_user_markdown(
    current_markdown: str,
    messages: list[dict],
    user_name: Optional[str] = None,
) -> str:
    prompt = _USER_PROMPT_TEMPLATE.format(
        user_name=user_name or "(匿名)",
        current_markdown=current_markdown.strip() or "(空)",
        conversation=_format_conversation(messages),
    )
    new_md = await _call_llm(prompt)
    if not new_md:
        return current_markdown
    # 防御：LLM 如果偷懒返回代码围栏/解释，剥离掉
    return _strip_code_fence(new_md)


async def refine_project_markdown(
    current_markdown: str,
    messages: list[dict],
    project_title: Optional[str] = None,
) -> str:
    prompt = _PROJECT_PROMPT_TEMPLATE.format(
        project_title=project_title or "本项目",
        current_markdown=current_markdown.strip() or "(空)",
        conversation=_format_conversation(messages),
    )
    new_md = await _call_llm(prompt)
    if not new_md:
        return current_markdown
    return _strip_code_fence(new_md)


def _strip_code_fence(text: str) -> str:
    t = text.strip()
    if t.startswith("```"):
        lines = t.split("\n")
        # 去掉第一行 ```xxx 和最后一行 ```
        if lines[-1].strip().startswith("```"):
            lines = lines[1:-1]
        else:
            lines = lines[1:]
        t = "\n".join(lines).strip()
    return t

File: client/src/data/agents/memoryMarkdownAgent.ts
/**
 * Memory Markdown Agent — 客户端版，从 backend `harness/agents/memory_markdown_agent.py` 移植。
 *
 * 责任：
 *   现有 .md + 新增对话/反馈/round signals → LLM → 增量改写后的 .md
 *
 * 两个 mode（对齐 backend）：
 * - user: 用户级（身份/职业/研究大方向）→ refineUser
 * - project: 项目级（本项目研究方向/子问题/关注点）→ refineProject
 *
 * Phase B B8 caller 视角的 refine() 抽象：
 *   接受 newSignals（feedback / conversation / round），统一拼接为对话块
 *   输入 LLM；返回 { updatedMd, summary }。
 */
import { llmManager as defaultLlmManager } from '../llm/manager'
import type { LLMResult } from '../llm/types'
import { loadPrompt } from './promptLoader'
import type { LLMGenerator } from './memoryAgent'

// ── 类型 ────────────────────────────────────────────────────────────────

/** 单条新信号 —— 来自 feedback / conversation / round 的统一形态。 */
export interface MemorySignal {
  /** 信号来源：feedback（4-bucket 反馈摘要）/ conversation（用户/AI 对话）/ round（每轮检索摘要） */
  source: 'feedback' | 'conversation' | 'round'
  /** 信号正文（一行或多行；caller 自行裁剪） */
  content: string
  /** 角色（仅 conversation 模式有意义；不传默认 'user'） */
  role?: 'user' | 'assistant' | 'system'
}

export interface RefineParams {
  /** 当前 .md 全文（首次 = ''） */
  currentMd: string
  /** 新增信号 */
  newSignals: MemorySignal[]
  /** 提示模板用：用户名（user mode）或项目标题（project mode） */
  subject?: string
}

export interface RefineResult {
  /** 增量改写后的 .md（无新信息时 = currentMd 一字不改） */
  updatedMd: string
  /** caller 视角的更新摘要（无变化 → 'no change'） */
  summary: string
}

/** Conversation 类消息（向 backend `_format_conversation` 对齐）。 */
export interface ConversationMessage {
  role: 'user' | 'assistant' | 'system' | string
  content: string
}

// ── 工具 ────────────────────────────────────────────────────────────────

const ROLE_TAG: Record<string, string> = {
  user: '用户',
  assistant: 'AI',
  system: '系统',
}

/** 把 MemorySignal[] 渲染成 prompt 里的「对话块」字符串。
 *
 *  对齐 backend `_format_conversation`：每条 `[tag] content`，空行分隔；空对话占位 `(无对话)`。
 */
function _formatSignals(signals: MemorySignal[]): string {
  const lines: string[] = []
  for (const s of signals) {
    const text = (s.content || '').trim()
    if (!text) continue
    if (s.source === 'conversation') {
      const tag = ROLE_TAG[s.role || 'user'] || s.role || 'user'
      lines.push(`[${tag}] ${text}`)
    } else if (s.source === 'feedback') {
      lines.push(`[反馈] ${text}`)
    } else if (s.source === 'round') {
      lines.push(`[检索轮次] ${text}`)
    } else {
      lines.push(text)
    }
  }
  return lines.length > 0 ? lines.join('\n\n') : '(无对话)'
}

/** 直接传 ConversationMessage[]（backend 老入口）的渲染等价物。 */
function _formatConversation(messages: ConversationMessage[]): string {
  const lines: string[] = []
  for (const m of messages) {
    const text = (m.content || '').trim()
    if (!text) continue
    const tag = ROLE_TAG[m.role] || m.role || 'user'
    lines.push(`[${tag}] ${text}`)
  }
  return lines.length > 0 ? lines.join('\n\n') : '(无对话)'
}

/** 防御 LLM 偷懒返回 ```...``` 代码块围栏。 */
function _stripCodeFence(text: string): string {
  let t = text.trim()
  if (!t.startsWith('```')) return t
  const lines = t.split('\n')
  // 去掉首行 ```xxx
  lines.shift()
  // 去掉尾行 ``` （若有）
  if (lines.length > 0 && lines[lines.length - 1].trim().startsWith('```')) {
    lines.pop()
  }
  t = lines.join('\n').trim()
  return t
}

/** 简单 diff 摘要：检测 updatedMd 与 currentMd 是否有实质变化。 */
function _summarize(currentMd: string, updatedMd: string, signalCount: number): string {
  const a = currentMd.trim()
  const b = updatedMd.trim()
  if (a === b) return 'no change'
  const delta = b.length - a.length
  const sign = delta >= 0 ? '+' : ''
  return `refined from ${signalCount} signals (${sign}${delta} chars)`
}

// ── MemoryMarkdownAgent ─────────────────────────────────────────────────

export class MemoryMarkdownAgent {
  private llm: LLMGenerator

  constructor(llm?: LLMGenerator | null) {
    this.llm = (llm as LLMGenerator) ?? defaultLlmManager
  }

  /** 用户级 .md 增量提炼。LLM 不可用 / 失败 → 原样返回 currentMd（不抛）。 */
  async refineUser(params: RefineParams): Promise<RefineResult> {
    return this._refine('agents/memory_markdown_user', params, 'user_name')
  }

  /** 项目级 .md 增量提炼。 */
  async refineProject(params: RefineParams): Promise<RefineResult> {
    return this._refine('agents/memory_markdown_project', params, 'project_title')
  }

  /** 通用入口（默认走 project mode；caller 可显式选 mode）。 */
  async refine(
    params: RefineParams & { mode?: 'user' | 'project' },
  ): Promise<RefineResult> {
    const mode = params.mode ?? 'project'
    return mode === 'user' ? this.refineUser(params) : this.refineProject(params)
  }

  /** Phase B B8 任务规范方法名 —— 项目级 .md 增量提炼（spec 别名，等价 refineProject）。 */
  async refineProjectMemory(params: RefineParams): Promise<RefineResult> {
    return this.refineProject(params)
  }

  /** Phase B B8 任务规范方法名 —— 用户级 .md 增量提炼（spec 别名，等价 refineUser）。 */
  async refineUserMemory(params: RefineParams): Promise<RefineResult> {
    return this.refineUser(params)
  }

  /** 直接传 backend 风格的 messages[]（兼容老 caller）。 */
  async refineUserFromMessages(
    currentMd: string,
    messages: ConversationMessage[],
    userName?: string,
  ): Promise<RefineResult> {
    return this._refineRaw('agents/memory_markdown_user', currentMd, _formatConversation(messages), userName, 'user_name')
  }

  async refineProjectFromMessages(
    currentMd: string,
    messages: ConversationMessage[],
    projectTitle?: string,
  ): Promise<RefineResult> {
    return this._refineRaw(
      'agents/memory_markdown_project',
      currentMd,
      _formatConversation(messages),
      projectTitle,
      'project_title',
    )
  }

  // ── private ──

  private async _refine(
    promptName: string,
    params: RefineParams,
    subjectVar: 'user_name' | 'project_title',
  ): Promise<RefineResult> {
    const conversation = _formatSignals(params.newSignals || [])
    return this._refineRaw(promptName, params.currentMd, conversation, params.subject, subjectVar)
  }

  private async _refineRaw(
    promptName: string,
    currentMd: string,
    conversation: string,
    subject: string | undefined,
    subjectVar: 'user_name' | 'project_title',
  ): Promise<RefineResult> {
    let promptText: string
    try {
      const pf = loadPrompt(promptName)
      const renderVars: Record<string, string> = {
        current_markdown: currentMd.trim() || '(空)',
        conversation,
      }
      renderVars[subjectVar] = subject || (subjectVar === 'user_name' ? '(匿名)' : '本项目')
      promptText = pf.render(renderVars)
    } catch (e) {
      console.warn('[MemoryMarkdownAgent] prompt render failed:', e)
      return { updatedMd: currentMd, summary: 'no change (prompt render failed)' }
    }

    let raw: string | null = null
    try {
      const result = await this.llm.generate(promptText, {
        temperature: 0.1,
        max_tokens: 3000,
      })
      if (typeof result === 'string') {
        raw = result
      } else if (result && typeof (result as LLMResult).text === 'string') {
        raw = (result as LLMResult).text
      }
    } catch (e) {
      console.warn('[MemoryMarkdownAgent] LLM call failed:', e)
      return { updatedMd: currentMd, summary: 'no change (LLM error)' }
    }

    const text = (raw ?? '').trim()
    if (!text) {
      return { updatedMd: currentMd, summary: 'no change (empty LLM result)' }
    }

    const updatedMd = _stripCodeFence(text)
    const signalCount = (() => {
      // 估算：按双换行段落数（_formatSignals 用 \n\n 分隔）
      if (conversation === '(无对话)') return 0
      return conversation.split(/\n\n+/).filter(Boolean).length
    })()
    return {
      updatedMd,
      summary: _summarize(currentMd, updatedMd, signalCount),
    }
  }
}

// ── 单例便捷入口 ─────────────────────────────────────────────────────────

let _singleton: MemoryMarkdownAgent | null = null
export function getMemoryMarkdownAgent(): MemoryMarkdownAgent {
  if (!_singleton) _singleton = new MemoryMarkdownAgent()
  return _singleton
}

export function _resetMemoryMarkdownAgentForTesting(
  replacement: MemoryMarkdownAgent | null = null,
): void {
  _singleton = replacement
}
